File: examples/guests/js/encoder/encoder.js
import { setEncoderFactory } from './generated/guest/contracts';
import { polyplug_init } from './generated/guest/init';
import { toStr } from '../../../../sdks/js/guest/polyplug_guest.js';

setEncoderFactory((bridge, hostLo, hostHi) => ({
    fn0: (input) => {
        const s = toStr(bridge, input);
        const data = s.startsWith("TRANSFORMED:") ? s.slice("TRANSFORMED:".length) : s;
        return data.replace(/\|/g, ",");
    }
}));

export { polyplug_init };

File: examples/guests/python/encoder/encoder.py
from generated.guest.contracts import (
    ENCODERPipelineEncoderPlugin,
    set_encoder_factory,
    polyplug_init,
)


class EncoderImpl(ENCODERPipelineEncoderPlugin):
    """The factory receives the HostApi pointer at polyplug_init time."""

    def __init__(self, host_ptr: int) -> None:
        # Host handle for this runtime, captured at construction.
        self._host_ptr: int = host_ptr

    def encode(self, input: str) -> str:
        s = input.removeprefix("TRANSFORMED:")
        return s.replace("|", ",")


# Register the factory; the generated polyplug_init constructs the
# implementation with its owning runtime's host pointer.
set_encoder_factory(EncoderImpl)
